ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.engine.memory;
+
+import static org.opensearch.ml.common.CommonValue.ML_MEMORY_MESSAGE_INDEX;
+import static org.opensearch.ml.common.CommonValue.ML_MEMORY_META_INDEX;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+
+import org.opensearch.action.get.GetRequest;
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.client.Client;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.Strings;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.ml.common.spi.memory.Memory;
+import org.opensearch.ml.common.spi.memory.Message;
+import org.opensearch.ml.engine.indices.MLIndicesHandler;
+import org.opensearch.ml.memory.action.conversation.CreateConversationResponse;
+import org.opensearch.ml.memory.action.conversation.CreateInteractionResponse;
+import org.opensearch.ml.repackage.com.google.common.collect.ImmutableMap;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.sort.SortOrder;
+
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Getter
+public class ConversationIndexMemory implements Memory {
+    public static final String TYPE = "conversation_index";
+    public static final String SESSION_ID = "session_id";
```

Comment:
Should we use `session_id` or `memory_id` ?

Reply:
https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/conversation/ConversationalIndexConstants.java#L69

> public final static String INTERACTIONS_CONVERSATION_ID_FIELD = "conversation_id";

Reply:
changed to conversation_id